chemprop/models/ffn.py

```diff
@@ -0,0 +1,355 @@
+                        build_ffn(
+                            first_linear_dim=features_size,
+                            hidden_size=hidden_size,
+                            num_layers=num_layers - 1,
+                            output_size=hidden_size,
+                            dropout=dropout,
+                            activation=activation,
+                        ),
+                        activation,
+                    )
+                else:
+                    self.ffn = nn.Identity()
+            self.ffn_readout = build_ffn(
+                first_linear_dim=base_output_size,
+                hidden_size=hidden_size,
+                num_layers=1,
+                output_size=output_size,
+                dropout=dropout,
+                activation=activation,
+            )
+            self.weights_readout = build_ffn(
+                first_linear_dim=base_output_size,
+                hidden_size=hidden_size,
+                output_size=1,
+                num_layers=weights_ffn_num_layers,
+                dropout=dropout,
+                activation=activation,
+            )
+        else:
+            if shared_ffn:
+                self.ffn_readout = nn.Sequential(
+                    ffn_base,
+                    build_ffn(
+                        first_linear_dim=base_output_size,
+                        hidden_size=hidden_size,
+                        num_layers=1,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                    ),
+                )
+            else:
+                self.ffn_readout = build_ffn(
+                    first_linear_dim=features_size,
+                    hidden_size=hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                )
+        self.constraint = constraint
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.tensor, List, torch.tensor, List, torch.tensor],
+        constraints: torch.tensor,
+        bond_types: torch.tensor,
+    ) -> torch.tensor:
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
```

Comment:
what happens if `self.ffn_type` is neither?

Reply:
`ffn_type` can only be `atom` or `bond`. The `FFNAtten` class will be constructed in the initialization of `MultiReadout()` class so the ffn_type is already given inherently. However, some users might want to use `FFNAtten` class directly for some purposes, so I will add `ValueError` here if other values are specified.